arbor/include/arbor/simple_sampler.hpp

```diff
@@ -22,16 +22,76 @@ struct trace_entry {
     V v;
 };
 
-// Trace data wraps a std::vector of trace_entry, optionally with
-// a copy of the metadata associated with a probe.
+// `trace_data` wraps a std::vector of trace_entry with
+// a copy of the probe-specific metadata associated with a probe.
+//
+// If `Meta` is void, ignore any metadta.
 
 template <typename V, typename Meta = void>
-struct trace_data: public std::vector<trace_entry<V>> {
-    util::optional<Meta> metadata;
+struct trace_data: private std::vector<trace_entry<V>> {
+    Meta meta;
+    explicit operator bool() const { return !this->empty(); }
+
+    using base = std::vector<trace_entry<V>>;
+    using base::size;
+    using base::empty;
+    using base::at;
+    using base::begin;
+    using base::end;
+    using base::clear;
+    using base::resize;
+    using base::push_back;
+    using base::emplace_back;
+    using base::operator[];
 };
 
 template <typename V>
-struct trace_data<V, void>: public std::vector<trace_entry<V>> {};
+struct trace_data<V, void>: private std::vector<trace_entry<V>> {
+    explicit operator bool() const { return !this->empty(); }
+
+    using base = std::vector<trace_entry<V>>;
+    using base::size;
+    using base::empty;
+    using base::at;
+    using base::begin;
+    using base::end;
+    using base::clear;
+    using base::resize;
+    using base::push_back;
+    using base::emplace_back;
+    using base::operator[];
+};
+
+// `trace_vector` wraps a vector of `trace_data`.
+//
+// When there are multiple probes associated with a probe id,
+// the ith element will correspond to the sample data obtained
+// from the probe with index i.
+//
+// The overriden method `at(i)` returns a const reference to the ith
+// element if it exists, or else to an empty trace_data value.
+
+template <typename V, typename Meta = void>
+struct trace_vector: private std::vector<trace_data<V, Meta>> {
+    const trace_data<V, Meta>& at(std::size_t i) const {
+        return i<this->size()? (*this)[i]: empty_trace;
```

Comment:
Why return an empty trace instead of throwing an exception?
`::at` methods in STL are expected to throw on an out of bounds check, so this might be a bit misleading.

Reply:
Yeah, this is a good point. I chose `at` because it is nice and short, and I don't want to make it a chore to access a trace. On the other hand, if I know a locset has 3 locations, and I want to get the trace from the 3rd one, I don't want things to blow up if it turns out there were no samples taken for that location.

Can you think of a better name for the method?

Reply:
Okay, let's use `get` (pers. comm.)